src/coordinate_search.jl

```diff
@@ -56,17 +62,20 @@ function coordinate_search(nlp :: AbstractNLPModel;
     k += 1
     el_time = time() - start_time
     tired = neval_obj(nlp) > max_eval >= 0 || el_time > max_time
+    ftol_reached = worst_fi - best_f <= ftol

```

Comment:
Just to make sure, is this really what you mean here? Do you want to stop when f stops decreasing significantly, or when f has decreased a certain amount?

Reply:
The general idea of `ftol` is to stop optimization when the target is insensitive to parameters changes. 
The `worst_fi - best_f` is the maximal amplitude of `f` inside the current local search. If the amplitude is too low then STOP.

In other words, my goal was to stop when f decreasing too insignificantly.

Reply:
In this kind of grid search, it is the mesh size (alpha) that guides the optimization. It could very well be that no point in the current local search improves f, but that doesn’t mean that you’re near a local solution. You will only know by reducing alpha and performing finer local searches.

This isn’t a criticism. I’m just trying to understand what the rationale for your changes is. You might obtain the desired effect by setting a “large” minimum value for alpha?!

Reply:
There are different applications of the optimization algorithms. Sometimes we are not interested in finding the parameters' values with the exact precise `tol`. In many cases, we need to estimate the minimal `f` value itself with some precision.
It is impossible to control the precision of `min(f)` in the current version.

Just an example. Before you start an optimization you never know the optimal value and scales of `x` and `f`. What if your target function is something like this: `f(x) = 1 + 1e9*(x[1] - 2)^2` (of course in many cases it is an implicit function and you cannot analyze it). If you run the optimizer with `tol=1e-3` you can expect the precision of `|dx[1]| < 1e-3` but at the same time precision `|df| < 1e+3` which is not enough for some goals. Implementing `ftol` allows you to control `x` and `f` tolerances depending on your needs.

It is also possible to create another example when the lack of `ftol` results in unnecessary multiple iterations to reach low `tol` having the necessary `min(f)` value.

To show that the `ftol` is not my unique want but a common problem I give a couple of references to optimizers that utilizing it:
- https://nlopt.readthedocs.io/en/latest/NLopt_Reference/#return-values
- https://julianlsolvers.github.io/Optim.jl/stable/#user/config/

Reply:
Just to clarify. I guess the problem of `ftol` is important not only for the "coordinate search" methods but for all derivative-free and non-derivative-free optimizers. By this request, I want to initiate the discussion about the additional stop criteria.
I suppose the criterion `stopval` or similar is also required in some cases. And it easy to implement with the corresponding return code. `stopval` is the threshold value of `f` which stops the optimization.

Multiple stop criteria are a good approach. They may be inactive by default but can be very helpful in some specific cases.

Reply:
I started a discussion here: https://github.com/JuliaSmoothOptimizers/Organization/discussions/22
It might be a while until we get the solver revamp completed, so let me know if this PR is needed sooner so I can properly review it.

Reply:
Thanks for your reply. My interest is strategic, not urgent.
Let me know if some help is required: review, implementation, test cases.

Reply:
I’m not questioning the usefulness of having several stopping conditions, or of the ftol condition. I have the impression that the ftol condition isn’t particularly useful in grid-based methods specifically.

Reply:
> I have the impression that the ftol condition isn’t particularly useful in grid-based methods specifically.
I guess it is. We've been developing the package https://github.com/insysbio/LikelihoodProfiler.jl which is based on derivative-free methods and ftol is our way to stop. Currently, we are using NM method from NLOpt but we are interested in other methods as well.